presto-main/src/main/java/io/prestosql/sql/planner/LiteralEncoder.java

```diff
@@ -210,6 +218,32 @@ public Expression toExpression(Object object, Type type)
             return new GenericLiteral("DATE", new SqlDate(toIntExact((Long) object)).toString());
         }
 
+        if (type instanceof TimestampType) {
+            TimestampType timestampType = (TimestampType) type;
+            String representation;
+            if (timestampType.isShort()) {
+                representation = TimestampToVarcharCast.cast(timestampType.getPrecision(), (Long) object).toStringUtf8();
+            }
+            else {
+                representation = TimestampToVarcharCast.cast(timestampType.getPrecision(), (LongTimestamp) object).toStringUtf8();
+            }
+            return new TimestampLiteral(representation);
+        }
+
+        if (type instanceof TimestampWithTimeZoneType) {
+            TimestampWithTimeZoneType timestampWithTimeZoneType = (TimestampWithTimeZoneType) type;
+            String representation;
+            if (timestampWithTimeZoneType.isShort()) {
+                representation = TimestampWithTimeZoneToVarcharCast.cast(timestampWithTimeZoneType.getPrecision(), (long) object).toStringUtf8();
+            }
+            else {
+                representation = TimestampWithTimeZoneToVarcharCast.cast(timestampWithTimeZoneType.getPrecision(), (LongTimestampWithTimeZone) object).toStringUtf8();
+            }
+            if (object.equals(parseTimestampWithTimeZone(timestampWithTimeZoneType.getPrecision(), representation))) {
```

Comment:
A comment will be helpful for why this check is needed. From what I understood this is to make sure we provide a TimestampLiteral only for cases where roundtrip is possible.

Reply:
This is weird check. Can't we determine if representation of given value as literal is possible just by looking at value itself. Not reverse-probing generated literal value?

Reply:
The check is necessary & is test covered by these cases in TLE

```
// DST change backward - no direct representation
assertRoundTrip(packDateTimeWithZone(1603589478_000L, TimeZoneKey.getTimeZoneKey("Europe/Warsaw")), createTimestampWithTimeZoneType(0), Long::equals);
assertRoundTrip(packDateTimeWithZone(1603589478_123L, TimeZoneKey.getTimeZoneKey("Europe/Warsaw")), createTimestampWithTimeZoneType(3), Long::equals);
assertRoundTrip(LongTimestampWithTimeZone.fromEpochMillisAndFraction(1603589478_123L, 123000000, TimeZoneKey.getTimeZoneKey("Europe/Warsaw")), createTimestampWithTimeZoneType(6), LongTimestampWithTimeZone::equals);
assertRoundTrip(LongTimestampWithTimeZone.fromEpochMillisAndFraction(1603589478_123L, 123456000, TimeZoneKey.getTimeZoneKey("Europe/Warsaw")), createTimestampWithTimeZoneType(9), LongTimestampWithTimeZone::equals);
assertRoundTrip(LongTimestampWithTimeZone.fromEpochMillisAndFraction(1603589478_123L, 123456789, TimeZoneKey.getTimeZoneKey("Europe/Warsaw")), createTimestampWithTimeZoneType(12), LongTimestampWithTimeZone::equals);
```

Reply:
> We may want to disallow that in general.

please file an issue for this, i'll be ref it from the TODO comment

Reply:
> The question is how do we ever end up with a timestamp w/ tz that falls in the gap or that has dual representation.

FWIW the Java Time solves this by encoding "point in time", zone _and_ offset within a value

```java
Instant instant = Instant.parse("2020-10-25T01:01:00Z");
ZoneId zone = ZoneId.of("Europe/Warsaw");
ZonedDateTime zonedDateTime = instant.atZone(zone);
System.out.printf("%s %s\n", zonedDateTime, zonedDateTime.toEpochSecond()); // 2020-10-25T02:01+01:00[Europe/Warsaw] 1603587660
zonedDateTime = zonedDateTime.toLocalDateTime().atZone(zone);
System.out.printf("%s %s\n", zonedDateTime, zonedDateTime.toEpochSecond()); // 2020-10-25T02:01+02:00[Europe/Warsaw] 1603584060
```

thus it can distinguish otherwise ambiguous values within DST backwards change.